src/kbc_modules/online_sev_kbc/mod.rs

```diff
@@ -0,0 +1,209 @@
+    // to aes-gcm, OpenSSL will only allow GCM when using AEAD. Because authentication is not
+    // handled here, AEAD cannot be used, therefore, CTR is used instead.
+    [(String::from("aes_256_ctr"), Cipher::aes_256_ctr())]
+        .iter()
+        .cloned()
+        .collect()
+}
+
+#[async_trait]
+impl KbcInterface for OnlineSevKbc {
+    fn check(&self) -> Result<KbcCheckInfo> {
+        Ok(KbcCheckInfo {
+            kbs_info: self.kbs_info.clone(),
+        })
+    }
+
+    async fn decrypt_payload(&mut self, annotation: &str) -> Result<Vec<u8>> {
+        let annotation_packet: AnnotationPacket = serde_json::from_str(annotation)
+            .map_err(|e| anyhow!("Failed to parse annotation: {}", e))?;
+
+        let key = self.get_key_from_kbs(annotation_packet.kid).await?;
+
+        let iv = decode(annotation_packet.iv).map_err(|e| anyhow!("Failed to decode IV: {}", e))?;
+        let wrapped_data = decode(annotation_packet.wrapped_data)
+            .map_err(|e| anyhow!("Failed to decode wrapped key: {}", e))?;
+        let wrap_type = annotation_packet.wrap_type;
+
+        let cipher = self
+            .ciphers
+            .get(&wrap_type)
+            .ok_or_else(|| anyhow!("Received unknown wrap type: {}", wrap_type))?;
+        // Redact decryption errors to avoid oracles
+        decrypt(*cipher, &key, Some(&iv), &wrapped_data).map_err(|_| anyhow!("Failed to decrypt"))
+    }
+}
+
+impl OnlineSevKbc {
+    #[allow(clippy::new_without_default)]
+    pub fn new(kbs_uri: String) -> OnlineSevKbc {
+        OnlineSevKbc {
+            kbs_info: HashMap::new(),
+            ciphers: get_ciphers(),
+            kbs_uri,
+            connection: load_connection(),
+        }
+    }
+
+    async fn query_kbs(&self, secret_type: String, secret_id: String) -> Result<Vec<u8>> {
+        let uri = format!("http://{}", self.kbs_uri).parse::<Uri>()?;
+
+        let channel = tonic::transport::Channel::builder(uri).connect_lazy();
+        let mut client = KeyBrokerServiceClient::new(channel);
+
+        let connection = self
+            .connection
+            .as_ref()
+            .map_err(|e| anyhow!("Failed to get injected connection. {}", e))?;
+        let guid = Uuid::new_v4().as_hyphenated().to_string();
+        let secret_request = RequestDetails {
+            guid: guid.clone(),
+            format: "binary".to_string(),
+            secret_type,
+            id: secret_id,
+        };
+
+        let request = tonic::Request::new(OnlineSecretRequest {
+            client_id: connection.client_id.as_hyphenated().to_string(),
+            secret_requests: vec![secret_request],
+        });
+
+        let response = client.get_online_secret(request).await?.into_inner();
+
+        let iv_bytes = base64::decode(response.iv)?;
+        let payload_bytes = base64::decode(response.payload)?;
+        let key_bytes = base64::decode(connection.key.clone())?;
```

Comment:
(future) It seems that all our crypto/keys code (in AA, AS, KBSs) would gain from marking such variables as `Secret` using the [secrecy crate](https://docs.rs/secrecy/latest/secrecy/). That's for a separate discussion/PR.